src/server/builder.ts

```diff
@@ -0,0 +1,417 @@
+                if (emitOutput.outputFiles && emitOutput.outputFiles.length > 0) {
+                    this.lastCheckedShapeSignature = this.computeHash(emitOutput.outputFiles[0].text);
+                }
+            }
+            return !lastSignature || this.lastCheckedShapeSignature !== lastSignature;
+        }
+    }
+
+    export interface Builder {
+        readonly project: Project;
+        getFilesAffectedBy(fileName: string): string[];
+        onProjectUpdateGraph(): void;
+        emitFile(fileName: string, writeFile: (path: string, data: string, writeByteOrderMark?: boolean) => void, forced?: boolean): boolean;
+    }
+
+    class NullBuilder implements Builder {
+        constructor(readonly project: Project) {
+        }
+
+        getFilesAffectedBy(fileName: string): string[] {
+            return undefined;
+        }
+
+        onProjectUpdateGraph() {
+        }
+
+        emitFile(fileName: string, writeFile: (path: string, data: string, writeByteOrderMark?: boolean) => void, forced?: boolean): boolean {
+            return true;
+        }
+    }
+
+    abstract class AbstractBuilder<T extends BuilderFileInfo> implements Builder {
+
+        private fileInfos: Map<T> = {};
+
+        constructor(public readonly project: Project) {
+        }
+
+        protected getFileInfo(fileName: string): T {
+            return this.fileInfos[fileName];
+        }
+
+        protected getAllFiles(): string[] {
+            return getKeys(this.fileInfos);
+        }
+
+        protected setFileInfo(fileName: string, info: T) {
+            this.fileInfos[fileName] = info;
+        }
+
+        protected removeFileInfo(fileName: string) {
+            if (hasProperty(this.fileInfos, fileName)) {
+                delete this.fileInfos[fileName];
+            }
+        }
+
+        protected forEachFileInfo(action: (fileInfo: T) => any) {
+            forEachValue(this.fileInfos, action);
+        }
+
+        abstract getFilesAffectedBy(fileName: string): string[];
+        abstract onProjectUpdateGraph(): void;
+
+        emitFile(fileName: string, writeFile: (path: string, data: string, writeByteOrderMark?: boolean) => void, forced?: boolean): boolean {
+            const fileInfo = this.getFileInfo(fileName);
+            if (!fileInfo) {
+                return false;
+            }
+
+            const { emitSkipped, outputFiles } = this.project.getFileEmitOutput(fileInfo.scriptInfo);
+            if (!emitSkipped) {
+                for (const outputFile of outputFiles) {
+                    writeFile(outputFile.name, outputFile.text, outputFile.writeByteOrderMark);
+                }
+            }
+            return !emitSkipped;
+        }
+    }
+
+    class NonModuleBuilder extends AbstractBuilder<BuilderFileInfo> {
+
+        onProjectUpdateGraph() {
+        }
+
+        getOrCreateFileInfo(fileName: string) {
+            if (!this.getFileInfo(fileName)) {

```

Comment:
it would rewrite to avoid multiple lookups
```ts
let fileInfo = this.getFileInfo(fileName);
if (!fileInfo) {
    fileInfo = new BuilderFileInfo(this.project.getScriptInfo(fileName), this.project);
    this.setFileInfo(fileName, fileInfo);
}
return fileInfo;
```